File: c++ version/1-100/24-swap-nodes-in-pairs.cpp
ListNode* swapPairs(ListNode* head) {
    ListNode* dummy = new ListNode(-1, head);
    ListNode* prev = dummy, *curr = head;
    while (curr && curr->next) {
        prev->next = curr->next;
        curr->next = curr->next->next;
        prev->next->next = curr;
        prev = curr;
        curr = curr->next;
    }
    ListNode* newHead = dummy->next;
    delete dummy;
    return newHead;
}

File: python version/24-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next: return head
        dummy_head = ListNode()
        dummy_head.next = head
        count = 0
        prev, curr = dummy_head, head
        while curr:
            count += 1
            if count == 2:
                tmp = curr.next
                curr.next = prev.next
                prev.next = curr
                curr.next.next = tmp
                prev = curr.next
                curr = tmp
                count = 0
            else:
                curr = curr.next
        
        return dummy_head.next